src/relay/transforms/fp32_to_fp16.cc

```diff
@@ -0,0 +1,337 @@
+    if (const TensorTypeNode* tensor_type = expr_type.as<TensorTypeNode>()) {
+      return CachedCast(expr, tensor_type->dtype, wanted_dtype);
+    } else if (const TupleTypeNode* tuple_type = expr_type.as<TupleTypeNode>()) {
+      Array<Expr> new_expr;
+      bool all_same = true;
+      for (size_t i = 0; i < (tuple_type->fields).size(); i++) {
+        Expr tuple_element = GetField(expr, i);
+        Type tuple_element_dtype = (tuple_type->fields)[i];
+        Expr casted_element = CastArg(tuple_element, tuple_element_dtype, wanted_dtype);
+        new_expr.push_back(casted_element);
+        all_same &= casted_element.same_as(tuple_element);
+      }
+      return all_same ? expr : Tuple(new_expr);
+    } else {
+      LOG(FATAL) << "Unsupported type " << expr_type << " we don't know how to cast for arguments!";
+      return expr;
+    }
+  }
+
+  std::pair<Array<Expr>, Array<Type>> CastAllArgs(const Array<Expr>& cur_args,
+                                                  const Array<Type>& cur_arg_types,
+                                                  const DataType& wanted_dtype) {
+    Array<Expr> new_args;
+    Array<Type> new_arg_types;
+    for (size_t i = 0; i < cur_args.size(); i++) {
+      Expr cur_arg = cur_args[i];
+      Type cur_arg_type = cur_arg_types[i];
+      Expr new_arg = CastArg(cur_arg, cur_arg_type, wanted_dtype);
+      Type new_arg_type = GetType(new_arg);
+      new_args.push_back(new_arg);
+      new_arg_types.push_back(new_arg_type);
+    }
+    return {new_args, new_arg_types};
+  }
+
+ public:
+  explicit AmpGraphCreator(ColorFunc colorer, OutputDtypeFunc output_dtype_func)
+      : ExprMutator(), colorer(colorer), output_dtype_func(output_dtype_func) {}
+
+  Expr VisitExpr_(const CallNode* call_node) {
+    FP16ConversionCategory initial_color = colorer(call_node);
+    auto new_op = this->Mutate(call_node->op);
+
+    // Mutate arguments to FP16 form first if possible and keep track of whether all floating point
+    // tensors are in FP16 form already. This is useful for propagating color.
+    Array<Expr> new_args;
+    Array<Type> new_arg_types;
+    bool all_args_fp16_compatible = true;
+    for (Expr arg : call_node->args) {
+      Expr new_arg = this->Mutate(arg);
+      Type new_arg_type = GetType(new_arg);
+      new_args.push_back(new_arg);
+      new_arg_types.push_back(new_arg_type);
+
+      if (all_args_fp16_compatible) {
+        // We can cast Vars and Constants to the right types so don't care about the types.
+        bool is_fp16_compatible = IsFP16Type(new_arg_type, true) || arg->IsInstance<VarNode>() ||
+                                  arg->IsInstance<ConstantNode>();
+        all_args_fp16_compatible &= is_fp16_compatible;
+      }
+    }
+
+    // Determine the final color.
+    FP16ConversionCategory final_color;
+    if (initial_color == GRAY) {
+      final_color = all_args_fp16_compatible ? GREEN : RED;
```

Comment:
Can you provide an example of FP16 incompatible?

Reply:
A contrived example with concat.

We have two branches whose outputs are fed into concat.

The first branch has a RED operation and returns an FP32 tensor.
The second branch returns an FP16 tensor.

Now that I say this, it might be better to be a bit smarter about GRAY ops when we have heterogeneous floating point types coming in.

E.g. let's say we had a concat with 10 fp16 args and 1 fp32 arg. It would be wasteful to default convert everything to fp32 and set the color as RED in this case. 

I will change this so the number of fp16/fp32 args are taken into account. If there is a majority of fp16 or a tie we color GREEN else we color RED. Sound good?

Reply:
The workaround sounds fine to me. Again I'd suggest putting these op-specific heuristics to op attribute instead of this pass.

Reply:
On closer thought I will lead things as is since only some ops will benefit from the trick I described. In the future exposing this to op-attributes might be worthwhile but I cannot think of a major savings that comes from this.